### PR TITLE
perf(pie): optimize performance of pie series

### PR DESCRIPTION
**Pull request by wind108369**
## Brief Information

This pull request is in the type of:

- [ ] bug fixing
- [ ] new feature
- [x] others

### What does this PR do?

饼图现在的渲染有性能问题，getPercentWithPrecision是很消耗性能的方法，每一个数据项都需要执行一次，在10000个数据点的时候就能把页面卡死，可以总的执行一次再取数使用。

### Fixed issues

#17173 

## Document Info

One of the following should be checked.

- [x] This PR doesn't relate to document changes
- [ ] The document should be updated later
- [ ] The document changes have been made in apache/echarts-doc#xxx

### ZRender Changes

- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

N.A.

### Merging options

- [x] Please squash the commits into a single one when merging.

### Other information

   没有删掉getPercentWithPrecision，因为api里面有暴露给外包
   
   测试之前需要执行 prepublish

### CODE REVIEW

**Review by Ovilia — CHANGES_REQUESTED** [review]
Please first add a test case by `npm run mktest pie-percent` to serve as benchmark to see how much improvement this PR makes.

**Comment by wind108369** [comment]
1000个数据点，pie-percent.html中的数据，优化前
image

优化后
image

File: src/util/number.ts [hunk 1]
```diff
@@ -279,6 +279,63 @@ export function getPercentWithPrecision(valueList: number[], idx: number, precis
...
+ * Get a data of given precision, assuring the sum of percentages
+ * in valueList is 1.
+ * The largest remainer method is used.
+ * https://en.wikipedia.org/wiki/Largest_remainder_method
+ *
+ * @param valueList a list of all data
+ * @param precision integer number showing digits of precision
+ * @return {Array<number>} percent ranging from 0 to 100
+ */
+export function getPercentSeats(valueList: number[], precision: number): number[] {
```
Comment (pissang) [inline src/util/number.ts:292]: Seems getPercentSeats shares lots of logic with getPercentWithPrecision. Could we reuse these code?

File: test/pie-percent.html [hunk 2]
```diff
@@ -0,0 +1,6077 @@
...
+            'echarts',
+            // 'map/js/china',
+            // './data/nutrients.json'
+        ], function (echarts) {
+            var option;
+
+            option = {
+                series: {
+                    type: 'pie',
+                    data: [{
```
Comment (plainheart) [inline test/pie-percent.html:58]: The mocked data looks too huge. I think we can use a generator function to replace them.

File: src/util/number.ts [hunk 3]
```diff
@@ -228,16 +228,33 @@ export function getPixelPrecision(dataExtent: [number, number], pixelExtent: [nu
  * @param precision integer number showing digits of precision
  * @return percent ranging from 0 to 100
  */
-export function getPercentWithPrecision(valueList: number[], idx: number, precision: number): number {
+ export function getPercentWithPrecision(valueList: number[], idx: number, precision: number): number {
```
Comment (plainheart) [inline src/util/number.ts:231]: Unexpected space here.

**Review by plainheart — APPROVED** [review]
LGTM. The optimization for the performance of the pie series looks quite significant!

File: src/chart/pie/PieSeries.ts [hunk 4]
```diff
@@ -159,31 +161,25 @@ class PieSeriesModel extends SeriesModel<PieSeriesOption> {
...
-            valueList.push(value);
-        });
-
-        params.percent = getPercentWithPrecision(
-            valueList,
-            dataIndex,
-            data.hostModel.get('percentPrecision')
-        );
-
+        params.percent = this.seats[dataIndex] / 100;
```
Comment (plainheart) [inline src/chart/pie/PieSeries.ts:182]: Here we need to divide the seat by `percentPrecision` instead of fixed 100.

**Review by plainheart — CHANGES_REQUESTED** [review]

File: src/util/number.ts [hunk 5]
```diff
@@ -275,8 +292,7 @@ export function getPercentWithPrecision(valueList: number[], idx: number, precis
         remainder[maxId] = 0;
         ++currentSum;
     }
-
-    return seats[idx] / digits;
+    return seats;
```
Comment (plainheart) [inline src/util/number.ts:295]: Or we do the division in the `getPercentSeats` function.

File: src/chart/pie/PieSeries.ts [hunk 6]
```diff
@@ -159,31 +161,25 @@ class PieSeriesModel extends SeriesModel<PieSeriesOption> {
...
+        const data = createSeriesDataSimply(this, {
             coordDimensions: ['value'],
             encodeDefaulter: zrUtil.curry(makeSeriesEncodeForNameBased, this)
         });
+        const valueList:number[] = [];
+        data.each(data.mapDimension('value'), function (value: number) {
+            valueList.push(value);
+        });
+
+        this.seats = getPercentSeats(valueList, data.hostModel.get('percentPrecision')).map(item => item / 100);
```
Comment (plainheart) [inline src/chart/pie/PieSeries.ts:173]: We prefer using `zrUtil.map`.

File: src/util/number.ts (hunk 5 above)
Comment (plainheart) [inline src/util/number.ts:295]: Here I think it's better to map the seats to a new array in which each seat is divided by the `digits`. Note that `zrUtil.map` is preferred.

File: src/util/number.ts [hunk 7]
```diff
@@ -233,11 +233,28 @@ export function getPercentWithPrecision(valueList: number[], idx: number, precis
         return 0;
     }
 
+    const seats = getPercentSeats(valueList, precision);
+    const digits = Math.pow(10, precision);
+
+    return (seats[idx] || 0) / digits;
```
Comment (plainheart) [inline src/util/number.ts:239]: Now returing `seats[idx] || 0` is enough.

File: src/chart/pie/PieSeries.ts [hunk 8]
```diff
@@ -159,31 +161,27 @@ class PieSeriesModel extends SeriesModel<PieSeriesOption> {
...
+        const data = createSeriesDataSimply(this, {
             coordDimensions: ['value'],
             encodeDefaulter: zrUtil.curry(makeSeriesEncodeForNameBased, this)
         });
+        const valueList:number[] = [];
+        data.each(data.mapDimension('value'), function (value: number) {
+            valueList.push(value);
+        });
+
+        this.seats = zrUtil.map(
```
Comment (plainheart) [inline src/chart/pie/PieSeries.ts:173]: No need to map. Just return the seats directly.

**Review by plainheart — APPROVED** [review]
Thanks for your contribution!

**Review by Ovilia — APPROVED** [review]

**Pull request merged**